src/function.rs

```diff
@@ -59,6 +60,17 @@ impl Function {
   }
 }
 
+fn abs_path(ctx: &FunctionContext, path: &str) -> Result<String, String> {
+  let abs_path_unchecked = ctx.search.working_directory.join(path).lexiclean();
+  match abs_path_unchecked.to_str() {
+    Some(wd) => Ok(wd.to_string()),
+    None => Err(format!(
+      "Directory path is not valid unicode: {}",
+      abs_path_unchecked.to_string_lossy()
+    )),
+  }
```

Comment:
If there's an error, it will be due to the working directory not being valid unicode, so I think it's better if the error message only mentions the working directory, like so:

```suggestion
  Ok(
    Utf8Path::new(context.search.working_directory.to_str().ok_or_else(|| {
      format!(
        "Working directory is not valid unicode: {}",
        context.search.working_directory.display()
      )
    }))
    .join(path)
    .lexiclean()
    .to_string()
  )
```

Reply:
Is `LexiClean` implemented for `Utf8Path` or `Utf8PathBuf`?

Reply:
Ah, no, it isn't, good point. I'll add an implementation soon. For now you can do something like:

```rust
if context.search.working_directory.to_str().is_none() {
      return Err(format!(
        "Working directory is not valid unicode: {}",
        context.search.working_directory.display()
      ));
}

Ok(
    context.search.working_directory
    .join(path)
    .lexiclean()
    .to_str()
    .unwrap()
    .to_owned()
)
```

Reply:
Is there any concern worth having over the doubled-up call to `to_str()`?

Reply:
(I've updated the PR to use a match statement which reports the working directory as being incorrect if the final join is bad)

Reply:
Calling `to_str()` twice shouldn't be an issue. A `just` process is going to spend the vast vast majority of its time executing recipe shell commands. Rust is already fast, and parsing, lexing, and doing local memory allocations are orders of magnitude faster than I/O, so I/O (launching external processes, reading their output, etc.) is going to be the bottleneck. So things like calling `to_str` are unlikely to make any difference in the perceived speed of `just` invocations.

That all being said, the updated version which does `to_str` once, but then only includes the current directory in the error message is fine too.